mcux_os_timer: Handle counter overflow in Power Mode 3

mmahadevan108 opened the pull request:
The counter that is used in Power Mode 3 to track System time could overflow for large timeouts. Add logic to ignore wakeup events from the timer due to counter overflows.

mmahadevan108 committed:
drivers: mcux_os_timer: Handle counter overflow in Power Mode 3

The counter that is used in Power Mode 3 to track System time could
overflow for large timeouts.
Add an API that the power system could use to ignore wakeup events
from the timer.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
soc: rw612: Handle counter overflow in Power Mode 3

The RTC counter that is used in Power Mode 3 to track
System time could overflow for large timeouts.
Add code to catch wakeup events due to this overflow and re-enter
Power Mode 3.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

decsny commented on drivers/timer/mcux_os_timer.c:
```diff
@@ -194,17 +222,23 @@ void sys_clock_set_timeout(int32_t ticks, bool idle)
+				/* Convert to microseconds and round up to the next value */
+				timeout = (((timeout / CYC_PER_TICK) * CYC_PER_TICK) * CYC_PER_US);
+			}
 			if (mcux_lpc_ostick_set_counter_timeout(timeout) == 0) {
```
i dont understand something here, why do you pass this value as timeout which then gets converted to ticks, but above you are setting it to a value with unit of ticks, it seems like a bug

mmahadevan108 replied:
Converting os timer ticks to counter ticks

decsny replied:
these names are terribly confusing, this code is made harder to understand by every variable being called the same

decsny commented on drivers/timer/mcux_os_timer.c:
```diff
@@ -194,17 +222,23 @@ void sys_clock_set_timeout(int32_t ticks, bool idle)
+			if (wait_forever) {
+				timeout = UINT32_MAX;
+			} else if (counter_overflow) {
+				timeout = counter_remaining_ticks;
```
shouldn't this unit be wrong ? this timeout value gets converted again to ticks, I think it will be an incorrect calculation

mmahadevan108 commented on soc/nxp/rw/power.c:
```diff
@@ -143,31 +144,39 @@ static void restore_mpu_state(void)
 	POWER_ClearWakeupStatus(DT_IRQN(DT_NODELABEL(pin0)));
 	NVIC_ClearPendingIRQ(DT_IRQN(DT_NODELABEL(pin0)));
 	EnableIRQ(DT_IRQN(DT_NODELABEL(pin0)));
 	POWER_EnableWakeup(DT_IRQN(DT_NODELABEL(pin0)));
```
These registers do not lose state when exiting from PM Mode 3. The GPIO IOMUX registers need to be reconfigured.

mmahadevan108 commented on include/zephyr/drivers/timer/nxp_os_timer.h:
```diff
@@ -0,0 +1,32 @@
+ *               to sleep.
+ *         false if the timer wakeup event is as expected
+ */
+bool z_nxp_os_timer_ignore_timer_wakeup(void);
```
I am not very particular about this prefix and can drop it. I was following what is already done in other timer drivers as example.
https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/timer/nrf_rtc_timer.h

mmahadevan108 commented on include/zephyr/drivers/timer/nxp_os_timer.h:
```diff
@@ -0,0 +1,32 @@
+extern "C" {
+#endif
+
+/** @brief The timer may have overflowed and triggered a wakeup.
```
I am following what is already inside other Zephyr timer drivers and I don't see them available in the public domain.

mmahadevan108 committed:
drivers: timer: Move MAX_WAIT to header file

Move MAX_WAIT define to the system timer header file
so system timer drivers can use this define.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

decsny reviewed (approved):
refactor can be separate
